Upgrade the public app to Next v10 and new architectural patterns

jaredcwhite opened the pull request:
Part of the work for #1081 

**Cypress test issue on CircleCI**: see https://github.com/bloom-housing/bloom/pull/1087#issuecomment-844555411

dominikx96 commented on sites/public/pages/index.tsx:
```diff
@@ -16,57 +15,55 @@ interface IndexProps {
+    const response = await axios.get(process.env.listingServiceUrl)
+    listings = response.data
+  } catch (error) {
+    console.log(error)
```
I think `console.error` would be more appropriate for error handling :)

dominikx96 commented on sites/public/pages/listing/[id]/[slug].tsx:
```diff
@@ -0,0 +1,55 @@
+  )
+}
+
+export async function getStaticPaths(context: any) {
```
I think `any` will produce additional lining warnings, so can we add proper types (preferred option) or disable `eslint@typescript-eslint/no-explicit-any` in these lines?

dominikx96 commented on sites/public/src/translations.ts:
```diff
@@ -0,0 +1,15 @@
+  es: spanishTranslations,
+  zh: chineseTranslations,
+  vi: vietnameseTranslations,
+} as Record<string, any>
```
Here we can use language enum as a type.

dominikx96 commented on ui-components/src/navigation/LanguageNav.tsx:
```diff
@@ -24,7 +24,9 @@ const LanguageNav = ({ items }: LanguageNavProps) => {
                 key={item.prefix}
-                onClick={() => changeLanguage(item.prefix)}
+                onClick={() => {
+                  void router.push(router.asPath, router.asPath, { locale: item.prefix || "en" })
```
Also, language enum would be helpful :)

dominikx96 reviewed (changes requested):
@jaredcwhite just a few comments, more quality improvements than nextjs update, overall looks good! 👍

jaredcwhite commented:
Thanks @dominikx96 for the feedback. I'm a little fuzzy on what we would use a language enum for. The keys are coming from the Next.js config which comes from an ENV var, and any deployment with different vars would need code changed manually. Can you explain your thinking further? Thanks!

jaredcwhite commented:
I just realized I totally forgot to update the _listings_ page! On it…

jaredcwhite commented:
Also need to refactor out the use of the `lRoute` helper

jaredcwhite commented:
> Also need to refactor out the use of the `lRoute` helper…and `LocalizedLink`

So…this is most definitely going to be a **breaking change** 🚨 to ui-components, requiring an update of all Next apps. Not a problem per se, but we'd obviously need to be aware for the housingbayarea branches. A package version bump there will need the whole Next v10.1 upgrade migrated over.

emilyjablonski commented on sites/partners/pages/listings/applications/add.tsx:
```diff
@@ -1,8 +1,9 @@
 import React from "react"
 import Head from "next/head"
-import { MetaTags, PageHeader, SiteAlert, t } from "@bloom-housing/ui-components"
```
Why was MetaTags removed from ui-components?

jaredcwhite replied:
MetaTags was using the `next/head` dependency, so in order to remove Next from the component package we'll need to handle anything in the HTML `<head>` through Next in the site folders.

emilyjablonski commented on sites/public/lib/ApplicationConductor.ts:
```diff
@@ -205,13 +205,12 @@ export default class ApplicationConductor {
   }
 
   routeTo(url: string) {
-    void Router.push(url).then(() => window.scrollTo(0, 0))
```
What was the point of the scrollTo logic before?

jaredcwhite replied:
It was going to a new page but not scrolling up to the top of the page by default. But in Next v10 they added that functionality automatically (sort of bonkers it wasn't there from the get-go).

emilyjablonski commented on sites/public/next.config.js:
```diff
@@ -46,145 +45,15 @@ module.exports = withCSS(
           sassLoaderOptions: {
             additionalData: tailwindVars,
           },
-          // exportPathMap adapted from https://github.com/zeit/next.js/blob/canary/examples/with-static-export/next.config.js
```
Is all of this automatically generated now?

jaredcwhite replied:
Yep, that's the beauty of the updated system. No routing information needed in the config file whatsoever! 🤩

software-project replied:
@jaredcwhite will we be able to `hide` routes that we don't want live yet?

emilyjablonski commented on sites/public/layouts/application.tsx:
```diff
@@ -36,44 +36,43 @@ const Layout = (props) => {
-          <LocalizedLink href="/listings" className="navbar-item">
-            {t("nav.listings")}
-          </LocalizedLink>
+          <Link href="/listings">
```
Feels a little weird to see a <Link> with a `href` wrapping an `a` tag without one, but I don't necessarily have a suggestion

jaredcwhite replied:
That's just how Next's link component works — although because I had to add a special `LinkComponent` wrapper for pushing into the `NavigationContext` for the components layer, it's possible we might decide to use `LinkComponent` ourselves too. But that'd be a "non-standard" pattern as far as Next is concerned.

emilyjablonski commented:
Could you provide some additional context in the PR description / an overview of what was done here?

jaredcwhite commented:
I walked through some of the thought process in https://github.com/bloom-housing/bloom/issues/1081 but I'll try to write up a TL;DR here:

* Dropped our own locale logic for all routing/nav/links because that's now built into Next. More info here: https://nextjs.org/docs/advanced-features/i18n-routing
* Switched to a better file-based routing mechanism for the listing page: `listing/[id]/[slug].tsx` (and `listing/[id].tsx` redirects to the one with the slug too) and use Next's data fetching pattern for static pages
* Added a preview page that is SSR'd for dynamic look at latest API data (we'll want to lock this down a bit once we have the partners MVP further along for listings)
* Made the necessary changes to remove the Next dependency in ui-components
* Simplified some older pages to use functional components/hooks for consistency with rest of the codebase

seanmalbert commented on sites/public/pages/index.tsx:
```diff
@@ -1,99 +1,92 @@
+  )
+}
+
+export async function getStaticProps() {
+  let listings = []
+  try {
+    const response = await axios.get(process.env.listingServiceUrl)
+    listings = response.data
+  } catch (error) {
+    console.error(error)
   }
+
+  return { props: { listings } }
 }
```
This is pretty cool. For anyone else wondering what the purpose of this was: https://nextjs.org/learn/basics/data-fetching/implement-getstaticprops

seanmalbert commented on sites/public/pages/listing/[id].tsx:
```diff
@@ -0,0 +1,22 @@
+  let response
+
+  try {
+    response = await axios.get(`${process.env.backendApiBase}/listings/${context.params.id}`)
```
This is a general question about the use of axios. Will we want to pass headers to the server, and if so would it be better to create a reusable axios instance with those headers? Event if not, is there value in creating an instance for each API with a specified base URL like:
```
const listingApi = axios.create({
    baseUrl: `${process.env.backendApiBase}/listings/`,
   headers: {}
});
```
So then we can call `await listingApi(context.params.id)`.

Or at least set a global axios instance so we don't have to specify process.env.backendApiBase each time.

pbn4 replied:
Also we should for quite a bit of time be using an autogenerated frontend client (refer to how partners portal is e.g. fetching applications), no direct axios.get calls.

jaredcwhite replied:
I agree, let's refactor out the direct axios calls…not sure we should do it here since this PR is already getting massive.

seanmalbert reviewed (commented):
Hey @jaredcwhite ,
This is great work. I more or less wanted to review this so I can be on top of the changes. There's a lot here and I'm only about 1/3 of the way through looking at it. I added some comments/questions so far and I look forward to looking at the rest.

My biggest concern so far is just the size of this PR. You've clearly put a lot of work into this and I don't want the reviews to be less thorough because of the effort it takes to review this amount of work in one go. I understand this may have been an all or nothing approach, but for reflection at the very least, how do you think we could break this up into smaller digestible pieces?

jaredcwhite commented on sites/public/lib/hooks.ts:
```diff
@@ -1,15 +1,19 @@
+  return (queryParams: Record<string, string> = {}) => {
+    const redirectUrl =
+      typeof router.query.redirectUrl === "string" && isInternalLink(router.query.redirectUrl)
+        ? router.query.redirectUrl
+        : null
 
     return router.push({
-      pathname: redirectUrl && typeof redirectUrl === "string" ? redirectUrl : defaultPath,
+      pathname: redirectUrl || defaultPath,
```
Geez, why didn't I think of that. 😅

jaredcwhite commented on sites/public/pages/index.tsx:
```diff
@@ -1,99 +1,92 @@
-    })
+    }
   }
 
-  public setSiteAlertMessage = (message: string, alertType: string) => {
-    this.setState({
-      alertMessage: message,
-      alertType: alertType,
-    })
-  }
+  const [alertInfo, setAlertInfo] = useState(blankAlertInfo())
```
Hmm, I feel like I had a rationale for it at the time, but could be we could just define the object directly.

jaredcwhite commented:
@seanmalbert:
> how do you think we could break this up into smaller digestible pieces?

I wish we could! But because of the nature of all the changes involved I figured we could either do one painfully large PR or two or three smaller PRs that are interdependent and collectively even more painful to deal with. Honestly this feels to me like the last of the "big shoes to drop" for the architecture that we've had over the past number of months (previous rounds being to move/rename packages, adding backend types for the frontend, etc.)…so I'm crossing my fingers it'll be smoother sailing from here on out.

jaredcwhite commented:
Also worth noting even after this we *still* need to update the Partners portal to match, so I'm not quite sure if we can even merge this into master until the portal changes are also done… So maybe @dominikx96 would want to branch off of this branch? 🤔

seanmalbert commented on sites/public/src/ListingView.tsx:
```diff
@@ -0,0 +1,400 @@
+      </ListingDetails>
+    </article>
+  )
+}
```
This feels like it could be broken up into smaller components to be used by this. Not saying we need to do that now, per se, there's just a lot going on in this file.

jaredcwhite replied:
Agreed.

seanmalbert reviewed (commented):
Hey @jaredcwhite ,
Again, this looks good overall. Added a couple more comments, for things we may want to do in the future.

jaredcwhite commented on sites/public/pages/listings.tsx:
```diff
@@ -1,87 +1,85 @@
+      return nowTime > moment(listing.applicationDueDate)
+    })
+  } catch (error) {
+    console.error(error)
```
Yeah, I feel like there could be (and should be, if not already) some kind of best practice Next recommends for handling errors across an application. I'll look into it further.

jaredcwhite commented:
> @jaredcwhite when I tried to open one of existing listings I see an error:
> 
> ![Image 2021-05-07 at 1 21 49 PM](https://user-images.githubusercontent.com/16655908/117443323-470d1080-af38-11eb-85ba-c46cb463630e.png)

@dominikx96 Hmm, as far as I can tell there's no case where the `getStaticPaths` function is called without a context object. Can you make sure you've yarn installed and try again?

seanmalbert commented:
@jaredcwhite , @dominikx96 and @software-project , is there any reason why the translation header patch #1244 can't be started with a branch of this now?

jaredcwhite commented:
@seanmalbert Yeah I'm not planning to enhance any additional functionality at this point, just trying to this PR to feature parity with master so it's mergeable.

jaredcwhite commented:
Going to work through the testing issues now.

jaredcwhite commented:
All right – other than a couple of failing CircleCI tests which seem to be related to frontend/backend API connection (not sure how to resolve), we seem to be in good shape! Both public and partners building on Netlify, ui-components tests pass.

dominikx96 commented:
@jaredcwhite I noticed this issue few times using commands from the main `package.json` (dev:all, dev:app:public, dev:app:partners, build:app:public, build:app:partners) - but now it works for me as expected. It sounds like the issue comes randomly until the backend is not ready (?)

software-project commented on ui-components/src/actions/LocalizedLink.tsx:
```diff
@@ -1,33 +1,11 @@
-  }
+// Legacy use only, deprecated
+const LocalizedLink = (props: PropsWithChildren<LinkProps>) => {
+  const { LinkComponent } = useContext(NavigationContext)
```
Very nice :)

jaredcwhite commented:
@seanmalbert and I tried to iterate a bit on why Cypress is failing on CircleCI. Sean verified he's able to get all the tests to pass locally. I suspect at this point it has something to do with the backend not being loaded yet when the `yarn build:app:public` command it run in setup, and prior to Next v10 that sort of worked and now it falls down hard. We've decided to merge this in for now so it's not blocking PRs and then try to get the Cypress setup working again shortly.
